### ISSUE TITLE
[handoff] Keystone-Σ₀ PLT — verify forward (Stage 0) → train → integrate

### ISSUE THREAD

**Issue by alex-place** [alex-place/lantern-os#1647]
**Handoff for the GPU/training machine.** A self-contained, pull-and-train package for a *proprietary* Σ₀ coder landed on `master` (squash `4fb5d04`, PR #1645). It is bootstrapped from the Apache-2.0 LoopCoder-V2 Parallel Loop Transformer (arXiv:2510.24824) and we own the forward pass end-to-end. Decision record: **ADR-0011** (`docs/adr/0011-proprietary-sigma0-base-model.md`, *Proposed — needs Alex's approval*).

Everything here lives in **`models/keystone-sigma0-plt/`**.

## ⚠️ The one blocking fact

The forward pass (`modeling_keystone_plt.py`) is a **faithful hand-port** of the vendor's vLLM-fork PyTorch implementation, but it was authored **without a GPU or the weights — so it has never run.** Weights map 1:1 onto the module tree (standard Llama + one `plt_gate` per layer), so *loading* is low-risk; the **2-loop orchestration** is where parity must be proven. **Do not train until Stage 0 = `PASS`.**

## Pull + run (on the ≥24 GB box)

```bash
git checkout master && git pull
cd models/keystone-sigma0-plt
pip install torch --index-url https://download.pytorch.org/whl/cu121
pip install -r requirements.txt
python download_and_patch.py --out ./checkpoint      # ~18 GB Apache-2.0 weights, one time
python check_parity.py --model ./checkpoint --dtype bf16   # STAGE-0 GATE
```

## Stage 0 — verify the forward (BLOCKING)

`check_parity.py` checks three things and writes a verdict to `data/convergence/keystone-plt-parity-log.jsonl`:
1. **Weight-key match** — every checkpoint tensor maps (no missing/unexpected). Should be clean given the 1:1 layout.
2. **Smoke generation** — ≥2/3 prompts produce coherent code. Garbage ⇒ the forward is wrong.
3. **Faithful parity (definitive)** — dump reference logits from the vLLM fork once and pass `--ref ref_logits.pt`; acceptance is `top1_agree ≥ 0.99`.

**Acceptance to leave Stage 0:** key-match clean **and** `top1_agree ≥ 0.99` against a vLLM reference.

### If parity fails, it's almost certainly one of these 3 reconstructed boundaries
All isolated + commented in `modeling_keystone_plt.py`, all config-gated:
- [ ] **CLP shift** — `plt_clp_shift` (default `False`, matches the visible inference path). Paper docstring says `a·E + b·shift(H)`. Flip in `config.json` if needed.
- [ ] **Sliding-window edge** — implemented `0 ≤ t−s < 64`. Try `≤ 64` (off-by-one) next.
- [ ] **Per-loop norm placement** — shared `model.norm` after every non-last loop and after the last.

### Capturing the vLLM reference (needed for definitive parity)
On a box with `yxing-bj/vllm` (the vendor fork): run the model on a fixed `input_ids`, save `{"input_ids", "logits"}` (`[1,T,V]`) to `ref_logits.pt`, copy into `models/keystone-sigma0-plt/`, pass `--ref ref_logits.pt`.

## Then (each stage evidence-gated — none auto-promotes the model)

- [ ] **Stage 0 — Parity** (above) ✅ to proceed
- [ ] **Stage 1 — Fit** — 4-bit (`--dtype 4bit`) VRAM + tok/s on the target box → `data/convergence/`
- [ ] **Stage 2 — Train** — `python train_lora.py --model ./checkpoint --data <your.jsonl> --out ./adapter` (QLoRA, adapter-only / frozen base per ADR-0010). Swap `data/sample_sft.jsonl` for a real verified-experience corpus.
- [ ] **Stage 3 — Eval** — `eval_humaneval_chat.py` head-to-head vs Qwen2.5-Coder on-box. A **win** is the only thing that flips `verified:true` for the model in `apps/lantern-garage/lib/local-model-registry.js`.
- [ ] **Stage 4 — Serve + council** — OpenAI/Ollama-compatible endpoint (reuse the `ouro_serve.py` pattern); register as a Σ₀ council member behind the verify gate.
- [ ] **Stage 5 — CSF** — archive base + adapter checkpoints (content-addressed) in the one CSF archive.

## Guardrails (do not skip)
- **Do not flip `verified:true` in `local-model-registry.js` without a reproduced on-box eval win** (External Reality Rule; the grounding gate exists for exactly this).
- **Weight adjustment is adapter-only over a frozen base** (ADR-0010) — no base-weight retraining; benchmark is never the training target.
- Weights are gitignored (~18 GB) — never commit `checkpoint/` or `adapter/`.

## Pointers
- Decision: `docs/adr/0011-proprietary-sigma0-base-model.md`
- Package + runbook: `models/keystone-sigma0-plt/README.md`
- Port source: vendor fork `yxing-bj/vllm:vllm/model_executor/models/iquest_loopcoder.py`
- Prior probe (vendor path, refuted): `experiments/loopcoder_v2_4bit_probe.py`, `data/convergence/loopcoder-probe-log.jsonl`
- Merged via PR #1645

**Comment by alex-place** [comment]
## On-box validation — RTX 3070 (8 GB), 4-bit

Ran the package end-to-end here. **Both structural verification and the training pipeline pass**; faithful parity still pending (needs ≥24 GB + vLLM ref).

**Env:** RTX 3070 8 GB · `.venv-train` cu121 · torch 2.5.1 · transformers 4.57.6 · bitsandbytes 0.49.2 · peft 0.19.1

### Stage-0 structural — PASS
- **Weight-key match: `missing=0 unexpected=0`** — the hand-written module tree matches the Apache-2.0 checkpoint exactly. (Checkpoint is ~7.6 B params, 3 safetensors shards.)
- **Fits the box: 5.71 GB peak** at 4-bit nf4.
- **Generates correct Python**: `return s == s[::-1]` (palindrome), `if n == 0:` (fib).
- Next-token argmax `66542` = `▁` (SentencePiece indent marker) — correct behavior after `\n` in code, not a bug.

### Training pipeline (QLoRA, ADR-0010 adapter-only) — PASS
- LoRA attaches: **29,360,128 trainable / 7,645,109,808 total = 0.38%**, base frozen.
- 8 optimizer steps ran: **gradients flow** (grad_norm 0.6–7.1, finite), loss finite (toy 4-sample set, so loss values prove the loop runs, not that anything useful was learned).
- Forward+backward through the **2-loop PLT with gradient checkpointing works** (no crash).
- **Adapter saved**: `adapter_model.safetensors` 117 MB, targets q/k/v/o/gate/up/down.

### Fixes landed
PR #1668 (validated by this run): UTF-8 console guard + tokenizer→fast-tokenizer patch in `download_and_patch.py`; `token_type_ids` drop + fast next-token check in `check_parity.py`.

### Still required (unchanged) — the blocking gate stays open
- **Faithful parity** is NOT established. Coherent code on a 4-bit quant does not prove the 2-loop math is bit-faithful. Run `check_parity.py --dtype bf16 --ref ref_logits.pt` on a ≥24 GB box; acceptance `top1_agree ≥ 0.99`. The three documented boundaries (CLP shift / window edge / per-loop norm) are the suspects if it misses.
- Note: 4-bit load took ~10–18 min here (disk + quantize bound on 8 GB); a 24 GB box loading bf16 will be far quicker.

**Comment by alex-place** [comment]
Closing as superseded by #1743, which is the same GPU-box handoff (Stage-0 parity) plus the newer, more complete Stage 1-3 breakdown (Adaptive Loop Gate, eval gate, serve+wire) and current state (Stage-0 local smoke PASSED, KV-cache BUILD flip #1766, serving wired #1758). Assigning that one to @Mookman11 instead of duplicating tracking across two open issues.
